### ISSUE TITLE
Remaining Issues in Gradient-Based Search

### ISSUE THREAD

**Issue by kksnowy** [vehicle-lang/vehicle#1211]
A long list of all the remaining to do's that I can think of.

**Things which are not working/not correct:**

1. [x] Unable to lift quantifiers for the andGate specification. Even after the lifting compiler pass, the quantifiers are not at the top of the loss AST (there is a MulRatTensor on top of them). -- Matthew

2. [x] Unable to compile properties with nested quantifiers (i.e. that require quantifier lifting) or forall.

Whenever we lift quantifiers in a property e.g. ```(exists (x1 : Real) . 0 < x1 < 1) or (exists (x2 : Real) . 1 < x2 < 2)``` becomes ```exists x1 . exists 2. 0 < x1 < 1 or 1 < x2 < 2```, the latter cannot be compiled and complains that the variable ```x2``` is missing lower and upper bounds.

Whenever we have a forall in a property (even when we don't lift the foralls) e.g. ```(forall (x1 : Real) . 0 < x1 < 1) or (forall (x2 : Real) . 1 < x2 < 2)``` becomes ```(exists x1 . not (0 < x1 < 1)) or (exists x2 . not (1 < x2 < 2))```, the latter cannot be compiled and complains that the variable ```x1``` is missing upper bounds.

3. [x] Quantifiers are currently lifted over disjunction which changes the meaning of the property.

4. [x] Compiling properties which involve unblocking produces an "Unblocking evaluation results in unevaluable result" error. This happens with the andGate, bounded, monotonicity, windController specs.
	
5. [x] Trying to compile a loss function for search or training using VehicleLoss produces this error -- Matthew

<img width="799" height="196" alt="Image" src="https://github.com/user-attachments/assets/33abc7e1-90c2-4e43-bbb5-fa7e5b651417" /> 

**Things which are working but not the best/can potentially be improved:**

1. [ ] I have put my PGD algorithm for search in ```samplers.py``` but it is very similar to the algorithm in ```get_loss()```. The difference is my algorithm uses a round-robin approach to optimise multiple inputs in turn and it always minimises the loss function. Maybe we can make a generic algorithm that can be used by get_loss() and search? 

2. [x] Since we are looking for counter-examples by finding witnesses to the negated property, the loss pattern is decreasing when searching for a counter-example which is unintuitive.

3. [ ] The ```load_specification()``` entry function name in Python doesn't indicate that this is only for training.
	
4. [x] Gradient-based searches on a loss function can repeatedly produce the same set of inputs despite starting from different initial points (may just be the choice of neural network).
	

**Things which are outstanding/not yet implemented:**
1. [ ] Compile properties to just a Bool function (i.e. not using a differentiable logic translation) so they can be evaluated for a set of inputs to see if the inputs are a witness or counter-example.
	
2. [ ] Perform search where the bounds of quantified variables depend on each other e.g.
```exists x . 0 < x < 1 and (exists y . x < y < 1 and f(x) < y) ``` -- Matthew

3. [x] My quantifier lifting compiler pass will throw an error if a property contains both ```forall``` and ```exists```. However, I think it should be able to handle properties like ```(forall x. P(x)) and (exists x. Q(x))```. Detection of alternating quantifiers needs to be fixed.
	
4. [ ] Test performing a search for specifications containing parameters and datasets.
	
5. [ ] Find a way to perform a search with a neural network in an external file (e.g. .onnx file)? - currently, my implementation only works with a PyTorch network which I define in unit tests.
	
6. [ ] Implement the search in TensorFlow
	
7. [ ] Experiments to evaluate the search implementation e.g. measuring time taken to produce witnesses/adversarial examples, visualising the loss pattern during the search. Need to find non-trivial specifications/neural networks.
	
8. [ ] The quantifier lifting compiler pass is using only ```ToBoolValue``` instead of ```ToBoolTensorValue``` i.e. it cannot handle specifications containing ```foreach```.

**Comment by MatthewDaggitt** [comment]
> Unable to lift quantifiers for the andGate specification. Even after the lifting compiler pass, the quantifiers are not at the top of the loss AST (there is a MulRatTensor on top of them). -- Matthew

Okay I've had a look at this one. The issue is that the compilation of the quantified values to loss values, may itself produce a set of disjuncted searches. For example, in the `andGate` spec the domain of the quantified variables is itself a set of disjuncts, i.e. either $0 <= x1 < 0.5$ or $0.5 <= x1 < 1$. So it splits it up into two searches!

Luckily we can handle this extremely easily, as we can just fold these disjuncts into the tree of hard boolean and/or operations that we are going to create! 

To this end, I have created a small PR (#1216) that refactors the `compileQuantifiers` method to expose the disjuncts rather than translating them to loss values immediately. Therefore once that is merged in, when you get to actually translate the lifted quantified expressions (the boxes in our diagram on Tuesday) to loss values, then you should call `compileQuantifiers` to do so and then simply append the returned disjuncts to the bottom of the hard Boolean tree of searches that we need to do. Does that make sense?

**Comment by kksnowy** [comment]
> > Unable to lift quantifiers for the andGate specification. Even after the lifting compiler pass, the quantifiers are not at the top of the loss AST (there is a MulRatTensor on top of them). -- Matthew
> 
> Okay I've had a look at this one. The issue is that the compilation of the quantified values to loss values, may itself produce a set of disjuncted searches. For example, in the `andGate` spec the domain of the quantified variables is itself a set of disjuncts, i.e. either 0 <= x 1 < 0.5 or 0.5 <= x 1 < 1 . So it splits it up into two searches!
> 
> Luckily we can handle this extremely easily, as we can just fold these disjuncts into the tree of hard boolean and/or operations that we are going to create!
> 
> To this end, I have created a small PR ([#1216](https://github.com/vehicle-lang/vehicle/pull/1216)) that refactors the `compileQuantifiers` method to expose the disjuncts rather than translating them to loss values immediately. Therefore once that is merged in, when you get to actually translate the lifted quantified expressions (the boxes in our diagram on Tuesday) to loss values, then you should call `compileQuantifiers` to do so and then simply append the returned disjuncts to the bottom of the hard Boolean tree of searches that we need to do. Does that make sense?

Thanks so much! Yes, I think that makes sense.

**Comment by kksnowy** [comment]
@MatthewDaggitt I have added the remaining golden specs which cannot be compiled and their reason to the "Things which are not working/not correct" section (there are only two underlying causes so far).
